src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -0,0 +1,202 @@
+                    analyzedNullabilityMap,
+                    snapshotBuilderOpt: null);
+            }
+
+#if DEBUG
+            internal void VerifyNode(BoundNode node)
+            {
+                if (node.Kind == BoundKind.TypeExpression || node.WasCompilerGenerated)
+                {
+                    return;
+                }
+
+                Debug.Assert(_incrementalSnapshots.ContainsKey(node.Syntax.SpanStart), $"Did not find a snapshot for {node} `{node.Syntax}.`");
+                Debug.Assert(_walkerGlobalStates.Length > _incrementalSnapshots[node.Syntax.SpanStart].GlobalStateIndex, $"Did not find global state for {node} `{node.Syntax}`.");
+            }
+#endif
+
+            /// <summary>
+            /// Simple int comparer that orders in descending order, so that searching for a snapshot will find the
+            /// closest position before a given position.
+            /// </summary>
+            private sealed class DescendingIntComparer : IComparer<int>
+            {
+                public int Compare(int x, int y) => y.CompareTo(x);
+            }
+
+            internal sealed class Builder
+            {
+
+                private readonly ArrayBuilder<SharedWalkerState> _walkerStates = ArrayBuilder<SharedWalkerState>.GetInstance();
+                /// <summary>
+                /// Snapshots are kept in a dictionary of position -> snapshot at that position. These are stored in descending order.
+                /// </summary>
+                private readonly ImmutableSortedDictionary<int, Snapshot>.Builder _incrementalSnapshots = ImmutableSortedDictionary.CreateBuilder<int, Snapshot>(new DescendingIntComparer());
+                /// <summary>
+                /// Every walker is walking a specific symbol, and can potentially walk each symbol multiple times
+                /// to get to a stable state. Each of these symbols gets a single global state slot, which this
+                /// dictionary keeps track of. These slots correspond to indexes into <see cref="_walkerStates"/>.
+                /// </summary>
+                private readonly PooledDictionary<Symbol, int> _symbolToSlot = PooledDictionary<Symbol, int>.GetInstance();
+                private int _currentWalkerSlot = -1;
+
+                internal SnapshotManager ToManagerAndFree()
+                {
+                    Debug.Assert(_currentWalkerSlot == -1, "Attempting to finalize snapshots before all walks completed");
+                    Debug.Assert(_symbolToSlot.Count == _walkerStates.Count);
+                    Debug.Assert(_symbolToSlot.Count > 0);
+                    _symbolToSlot.Free();
+                    return new SnapshotManager(
+                        _walkerStates.ToImmutableAndFree(),
+                        _incrementalSnapshots.ToImmutable());
+                }
+
+                internal int EnterNewWalker(Symbol symbol)
+                {
+                    Debug.Assert(symbol is object);
```

Comment:
@AlekseyTs unfortunately the assertion that this incoming `Symbol` is related to the current symbol proved to not be true in bad code scenarios like this one: https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Semantic/Semantics/LookupTests.cs#L1540-L1555.